rust/counter | reimplement motoko/counter in rust

JamesMcGuigan opened the pull request:
**Overview**
Counter is the simplest state management canister example, but was previously only implemented in Motoko.

This pull request ports the Motoko example to Rust, to highlight the major syntax differences between languages.

**Additions**
I have added in rust unit tests that can be run via `cargo test`

**Observations**
In Motoko, Nat is considered a first-class primitive, and state can be managed as a class variable:

In Rust `Nat32` / `Nat64` is a more convenient primitive, as `Nat` maps to `num-bigint::BigUInt` which is a heavyweight heap vector that doesn't implement the Copy trait (triggering various borrow checker puzzles).

Additionally, global state requires using RefCell indirection:
```
thread_local! { static COUNTER: RefCell<Nat> = RefCell::new(Nat::from(0)); }
```

sesi200 commented on rust/counter/README.md:
```diff
@@ -0,0 +1,90 @@
+persistent `counter` variable to store an arbitrary precision natural number
+that represents the current value of the counter.
+
+By using the Motoko keyword `stable` when declaring the `counter` variable,
+the value of this variable will automatically be preserved whenever your canister code is
+upgraded. Without the `stable` keyword, a variable is deemed `flexible`, and its value
+is reinitialized on every canister upgrade, i.e. whenever new code is deployed to the canister.
+
+To learn more about these features of Motoko, see:
+* https://sdk.dfinity.org/docs/language-guide/motoko.html#_orthogonal_persistence
+* https://sdk.dfinity.org/docs/language-guide/upgrades.html#_declaring_stable_variables
```
Old stuff from the Motoko original, should be adapted for Rust.

JamesMcGuigan replied:
I have rewritten this section in the README attempting to discuss the mechanics of Rust. Please can you confirm that what I have said is technically correct.

sesi200 commented on rust/counter/README.md:
```diff
@@ -0,0 +1,90 @@
+Verify the following before running this demo:
+
+*  You have downloaded and installed the [DFINITY Canister
+   SDK](https://sdk.dfinity.org).
```
```suggestion
   SDK](https://internetcomputer.org/docs/current/developer-docs/build/install-upgrade-remove).
```
SDK has moved

JamesMcGuigan replied:
DONE + updated link in Motoko counter example

sesi200 commented on rust/counter/dfx.json:
```diff
@@ -0,0 +1,23 @@
+      "packtool": ""
+    }
+  },
+  "dfx": "0.11.1",
```
```suggestion
```
Can we remove this line, please? With it removed, it will always run the currently installed version; with the line, as soon as a new version is released, this will cause issues.

sesi200 reviewed (changes requested):
Thank you very much, @JamesMcGuigan! I have a few nitpicks I'd like to see cleaned up, otherwise this is great.

JamesMcGuigan commented:
@sesi200 Thank you for your feedback and code review. I have implemented all your suggestions and rewritten the documentation to describe Rust. 

Could you please confirm that what I have written is technically correct, if there is any important information missing, and if the writing style is suitable for the examples repo.

Thank you.

sesi200 commented on rust/counter/README.md:
```diff
@@ -1,36 +1,49 @@
+declare global state variable within a `thread_local!` block
+and wrapping `COUNTER` in a `RefCell` pointer.
+
+This is the equivalent of the `static` keyword in Mokoto, and the value of 
```
```suggestion
This is the equivalent of the `stable` keyword in Mokoto, and the value of 
```
Typo?

sesi200 reviewed (changes requested):
Awesome, thank you very much! I think there is one typo left, but I want to confirm with you that I this really is a typo. Otherwise I'd like to merge this.

JamesMcGuigan commented:
Looking at the Motoko counter example, the correct keyword is indeed `stable` and not `static`
```
actor Counter {
  stable var counter = 0;
}
```
- https://github.com/dfinity/examples/blob/master/motoko/counter/src/Main.mo

Which agrees with the documention
- https://github.com/dfinity/motoko/blob/master/design/Stable.md

Thank you for spotting this. I have committed your code review update.
